backends/imgui_impl_osx.mm

```diff
@@ -329,59 +364,45 @@ bool ImGui_ImplOSX_HandleEvent(NSEvent* event, NSView* view)
         return io.WantCaptureMouse;
     }
 
-    // FIXME: All the key handling is wrong and broken. Refer to GLFW's cocoa_init.mm and cocoa_window.mm.
-    if (event.type == NSEventTypeKeyDown)
+    if (event.type == NSEventTypeKeyDown || event.type == NSEventTypeKeyUp)
     {
-        NSString* str = [event characters];
-        NSUInteger len = [str length];
-        for (NSUInteger i = 0; i < len; i++)
+        unsigned short code = event.keyCode;
+        IM_ASSERT(code >= 0 && code < IM_ARRAYSIZE(io.KeysDown));
+        io.KeysDown[code] = event.type == NSEventTypeKeyDown;
+        NSEventModifierFlags flags = event.modifierFlags;
+        io.KeyCtrl  = (flags & NSEventModifierFlagControl) != 0;
+        io.KeyShift = (flags & NSEventModifierFlagShift) != 0;
+        io.KeyAlt   = (flags & NSEventModifierFlagOption) != 0;
+        io.KeySuper = (flags & NSEventModifierFlagCommand) != 0;
+
+        if (event.type == NSEventTypeKeyDown)
         {
-            int c = [str characterAtIndex:i];
-            if (!io.KeySuper && !(c >= 0xF700 && c <= 0xFFFF) && c != 127)
-                io.AddInputCharacter((unsigned int)c);
-
-            // We must reset in case we're pressing a sequence of special keys while keeping the command pressed
-            int key = mapCharacterToKey(c);
-            if (key != -1 && key < 256 && !io.KeySuper)
-                resetKeys();
-            if (key != -1)
-                io.KeysDown[key] = true;
+            NSString* str = [event characters];
+            io.AddInputCharactersUTF8(str.UTF8String);
```

Comment:
Some characters were previously intentionally filtered, why removing this?

Reply:
Ahh, sorry. I will provide a fix for this.

I will also resolve the OpenGL build

Reply:
It's worth double checking if those checks are still valid, or if they are specific to old versions of the OS. like the 127 code test I believe is checking for the delete key, but I don't have any keyboards that generate the 127 code. And I think the 0xF700 0xFF00 range might be an old way of filtering F keys? I might be wrong on this, but I think that's what this line of code was intended for.

Reply:
I've reworked it to use macOS's `NSTextInputClient`, which is the same as GLFW:

https://github.com/glfw/glfw/blob/b55a517ae0c7b5127dffa79a64f5406021bf9076/src/cocoa_window.m#L722-L723

and SDL2:

https://github.com/spurious/SDL-mirror/blob/e17aacbd09e65a4fd1e166621e011e581fb017a8/src/video/cocoa/SDL_cocoakeyboard.m#L53

This means we only send characters to ImGui when macOS tells us so:

https://github.com/stuartcarnie/imgui/blob/28c32a6c105441eb7557a3c58f9d3d3b69fedc38/backends/imgui_impl_osx.mm#L92-L107